pkg/controller/build/build_controller.go

```diff
@@ -312,114 +313,103 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileBuild) validateRuntime(runtime *build.Runtime) error {
-	if len(runtime.Paths) == 0 {
-		return fmt.Errorf("the property 'spec.runtime.paths' must not be empty")
+// UpdateBuildStatusAndRetreat returns an error if an update fails, this should force
+// a new reconcile until the API call succeeds. If return is nil, no further reconciliations
+// will take place
+func (r *ReconcileBuild) UpdateBuildStatusAndRetreat(ctx context.Context, b *build.Build) (reconcile.Result, error) {
+	if err := r.client.Status().Update(ctx, b); err != nil {
+		return reconcile.Result{}, err
 	}
-	return nil
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileBuild) validateRuntimeFailed(b *build.Build) bool {
+	if len(b.Spec.Runtime.Paths) == 0 {
+		MarkBuildStatus(b, build.RuntimePathsCanNotBeEmpty, fmt.Sprintf("the property 'spec.runtime.paths' must not be empty"))
+		return true
+	}
+	return false
 }
 
-func (r *ReconcileBuild) validateStrategyRef(ctx context.Context, s *build.StrategyRef, ns string) error {
-	if s.Kind != nil {
-		switch *s.Kind {
+func (r *ReconcileBuild) validateStrategyRef(ctx context.Context, b *build.Build) error {
+
+	if b.Spec.StrategyRef.Kind != nil {
+		switch *b.Spec.StrategyRef.Kind {
 		case build.NamespacedBuildStrategyKind:
-			if err := r.validateBuildStrategy(ctx, s.Name, ns); err != nil {
+			if err := r.validateBuildStrategy(ctx, b); err != nil {
 				return err
 			}
 		case build.ClusterBuildStrategyKind:
-			if err := r.validateClusterBuildStrategy(ctx, s.Name); err != nil {
+			if err := r.validateClusterBuildStrategy(ctx, b); err != nil {
 				return err
 			}
 		default:
-			return fmt.Errorf("unknown strategy %v", *s.Kind)
+			return fmt.Errorf("unknown strategy kind: %v", *b.Spec.StrategyRef.Kind)
 		}
 	} else {
 		ctxlog.Info(ctx, "buildStrategy kind is nil, use default NamespacedBuildStrategyKind")
-		if err := r.validateBuildStrategy(ctx, s.Name, ns); err != nil {
+		if err := r.validateBuildStrategy(ctx, b); err != nil {
 			return err
 		}
 	}
+
 	return nil
 }
 
-func (r *ReconcileBuild) validateBuildStrategy(ctx context.Context, n string, ns string) error {
+func (r *ReconcileBuild) validateBuildStrategy(ctx context.Context, b *build.Build) error {
 	list := &build.BuildStrategyList{}
 
-	if err := r.client.List(ctx, list, &client.ListOptions{Namespace: ns}); err != nil {
-		return errors.Wrapf(err, "listing BuildStrategies in ns %s failed", ns)
-	}
-
-	if len(list.Items) == 0 {
-		return errors.Errorf("none BuildStrategies found in namespace %s", ns)
+	if err := r.client.List(ctx, list, &client.ListOptions{Namespace: b.Namespace}); err != nil {
```

Comment:
+1 on eliminating the `none BuildStrategies found in namespace` error. Can we now change this `List` call into a `Get` call to directly try to retrieve the referenced BuildStrategy? Same applies to ClusterBuildStrategy related code.

Reply:
It's a good idea I think!

Reply:
@xiujuan95 can you pls address this one via https://github.com/qu1queee/build/pull/11 . Thanks in advance

Reply:
@qu1queee Done! I opened a new PR: https://github.com/qu1queee/build/pull/12 to address this, because [previous PR](https://github.com/qu1queee/build/pull/11) had too many conflicts, I will close it.